### ISSUE TITLE
Weird build failres with --isolation=nspawn and bootstrap image

### ISSUE THREAD

**Issue by FrostyX** [fedora-copr/copr#2906]
I am seeing these failures for the first time and don't really know what is going on. But it looks like there is some problem with bootstrap images. I have troubles isolating a reproducer, so I am creating just a Copr issue.

```
Finish(bootstrap): cleaning package manager metadata
INFO: Using bootstrap image: registry.fedoraproject.org/fedora:39
INFO: Pulling image: registry.fedoraproject.org/fedora:39
INFO: Copy content of container registry.fedoraproject.org/fedora:39 to /var/lib/mock/fedora-39-x86_64-bootstrap-1694347273.676351/root
INFO: mounting registry.fedoraproject.org/fedora:39 with podman image mount
INFO: image registry.fedoraproject.org/fedora:39 as /var/lib/containers/storage/overlay/8237bc5cdb06944d641a1fecb52270d275e570e4d2dd3502a60357279a745db9/merged
INFO: umounting image registry.fedoraproject.org/fedora:39 (/var/lib/containers/storage/overlay/8237bc5cdb06944d641a1fecb52270d275e570e4d2dd3502a60357279a745db9/merged) with podman image umount
INFO: Package manager dnf detected and used (fallback)
INFO: Bootstrap image not marked ready
Start(bootstrap): installing dnf tooling
Failed to mount /proc/sys (type n/a) on /proc/sys (MS_BIND ""): Invalid argument
Child died too early.
WARNING: DNF command failed, retrying, attempt #2, sleeping 10s
Failed to mount /proc/sys (type n/a) on /proc/sys (MS_BIND ""): Invalid argument
Failed to create /user.slice/user-1000.slice/session-12.scope/payload subcgroup: Structure needs cleaning
WARNING: DNF command failed, retrying, attempt #3, sleeping 10s
Failed to mount /proc/sys (type n/a) on /proc/sys (MS_BIND ""): Invalid argument
Failed to create /user.slice/user-1000.slice/session-12.scope/payload subcgroup: Structure needs cleaning
ERROR: Exception(/var/lib/copr-rpmbuild/workspace/workdir-50w_wl2q/copr-cli/copr-cli.spec) Config(fedora-39-x86_64) 0 minutes 27 seconds
INFO: Results and/or logs in: /var/lib/copr-rpmbuild/results
INFO: Cleaning up build root ('cleanup_on_failure=True')
Start: clean chroot
Finish: clean chroot
WARNING: Machine 1da35ef953f4477384576f18a7064cd7 still running. Killing...
ERROR: Command failed: 
 # /usr/bin/systemd-nspawn -q -M 50743cd0fe0a4142b9b2dbb2c5f8eea6 -D /var/lib/mock/fedora-39-x86_64-bootstrap-1694347273.676351/root -a --console=pipe --setenv=TERM=vt100 --setenv=SHELL=/bin/bash --setenv=HOME=/var/lib/mock/fedora-39-x86_64-bootstrap-1694347273.676351/root/installation-homedir --setenv=HOSTNAME=mock --setenv=PATH=/usr/bin:/bin:/usr/sbin:/sbin --setenv=PROMPT_COMMAND=printf "\033]0;<mock-chroot>\007" --setenv=PS1=<mock-chroot> \s-\v\$  --setenv=LANG=C.UTF-8 --setenv=LC_MESSAGES=C.UTF-8 --setenv=SYSTEMD_NSPAWN_TMPFS_TMP=0 --setenv=SYSTEMD_SECCOMP=0 --resolv-conf=off /usr/bin/dnf-3 --releasever 39 --setopt=deltarpm=False --setopt=allow_vendor_change=yes --allowerasing --disableplugin=local --disableplugin=spacewalk --disableplugin=versionlock install python3-dnf python3-dnf-plugins-core --setopt=tsflags=nocontexts --setopt=tsflags=nocontexts --setopt=tsflags=nocontexts
Failed to mount /proc/sys (type n/a) on /proc/sys (MS_BIND ""): Invalid argument
Failed to create /user.slice/user-1000.slice/session-12.scope/payload subcgroup: Structure needs cleaning

WARNING: Machine 3119827494fb4bf3974e06bd910d03c8 still running. Killing...
Copr build error: Build failed
```

In case somebody needs a quick workaround, you can go to your project settings and change "Mock bootstrap" from default to "Enable", it will work fine.

**Comment by FrostyX** [comment]
Will this be fixed by https://github.com/rpm-software-management/mock/issues/1204 ?

**Comment by praiskup** [comment]
If I read the output correctly, the image is correctly downloaded and used to generate the bootstrap chroot.  The problem happens later when trying to install 'python3-dnf-plugins-core' **into the image** using the **in-downloaded-image dnf**.  Mock expects that DNF is working, so....  not sure where are the roots of this problem... :(

**Comment by frenzymadness** [comment]
I was about to report the same bug. Thank you for the workaround! It works for me as well.

**Comment by praiskup** [comment]
This seems to be caused by a new systemd-nspawn from systemd-container-253.9-1.fc38.x86_64

**Comment by praiskup** [comment]
But indeed, when the bootstrap image feature is turned off, systemd-nspawn works fine.  So it seems that the `podman image mount` storage is not compatible with `systemd-nspawn` expectations.

**Comment by praiskup** [comment]
> Could it be this issue?

Thank you for the reference!  I'm not sure, doesn't look like it.  In this case, `systemd-nspawn` fails to even start the inner `dnf install` command.  This fails much earlier.

It's all weird; when I run the `systemd-nspawn` command manually, it just works:
```
/usr/bin/systemd-nspawn -q -M c63d6e1e9fb6454195d362d408241de6 -D /var/lib/mock/fedora-38-x86_64-bootstrap/root -a --console=pipe --setenv=TERM=vt100 --setenv=SHELL=/bin/bash --setenv=HOME=/var/lib/mock/fedora-38-x86_64-bootstrap/root/installation-homedir --setenv=HOSTNAME=mock --setenv=PATH=/usr/bin:/bin:/usr/sbin:/sbin '--setenv=PROMPT_COMMAND=printf "\033]0;<mock-chroot>\007"' '--setenv=PS1=<mock-chroot> \s-\v\$ ' --setenv=LANG=C.UTF-8 --setenv=LC_MESSAGES=C.UTF-8 --resolv-conf=off /usr/bin/dnf-3 --releasever 38 --setopt=deltarpm=False --setopt=allow_vendor_change=yes --allowerasing --disableplugin=local --disableplugin=spacewalk --disableplugin=versionlock install python3-dnf python3-dnf-plugins-core --setopt=tsflags=nocontexts
```

Mock drops some privileges though by default, I need to find the related logic....

Though the easiest way around is to use `--isolation=simple` ATM.

**Comment by praiskup** [comment]
Related update: https://bodhi.fedoraproject.org/updates/FEDORA-2023-13e3892b6e

**Comment by praiskup** [comment]
I wasn't able to reproduce properly.  This doesn't seem to be related to the process permission dropping Mock does; if I let Mock prepare the root directory (using tarball from mounted podmanimage), even a fully privileged unconfined root process fails to execute nspawn process inside...

For now I installed a work-around which doesn't update systemd on our builders: https://pagure.io/fedora-infra/ansible/c/ff74364720f1f7ce05faa5d1f2f3f970037d4671

**Comment by praiskup** [comment]
@keszybz doesn't this error look familiar to you?

> $ sudo /usr/bin/systemd-nspawn -q -M dfc902a886be4291b35c6e10510b46ef -D /var/lib/mock/fedora-38-x86_64-bootstrap/root -a --console=pipe --setenv=TERM=vt100 --setenv=SHELL=/bin/bash --setenv=HOME=/var/lib/mock/fedora-38-x86_64-bootstrap/root/installation-homedir --setenv=HOSTNAME=mock --setenv=PATH=/usr/bin:/bin:/usr/sbin:/sbin '--setenv=PROMPT_COMMAND=printf "\033]0;<mock-chroot>\007"' '--setenv=PS1=<mock-chroot> \s-\v\$ ' --setenv=LANG=C.UTF-8 --setenv=LC_MESSAGES=C.UTF-8 --resolv-conf=off /usr/bin/dnf-3 --releasever 38 --setopt=deltarpm=False --setopt=allow_vendor_change=yes --allowerasing --disableplugin=local --disableplugin=spacewalk --disableplugin=versionlock install python3-dnf python3-dnf-plugins-core --setopt=tsflags=nocontexts
> Failed to mount /proc/sys (type n/a) on /proc/sys (MS_BIND ""): Invalid argument

Might that be related to https://bugzilla.redhat.com/show_bug.cgi?id=2210335 as Florian suggests?

**Comment by praiskup** [comment]
> For now I installed a workaround which doesn't update system packages on our builders: https://pagure.io/fedora-infra/ansible/c/ff74364720f1f7ce05faa5d1f2f3f970037d4671

Just to underline this, **the workaround is in production**.  So Copr should build at this moment with the default setup.

**Comment by keszybz** [comment]
> Might that be related to https://bugzilla.redhat.com/show_bug.cgi?id=2210335 as Florian suggests?

I'm not sure. In [2210335](https://bugzilla.redhat.com/show_bug.cgi?id=2210335) we get EPERM. The kernel does not allow the mount to happen, so that is the right error code. The issue is that `/proc` is not fully visible, so the kernel refuses a mount that would expose the hidden parts. In the report here, the error is EINVAL. Also we're trying to mount `/proc/sys`, which is bind-mounted by `systemd-nspawn`, so it should not be a problem to mount it again… Overall, I'm not sure what is going on here.

**Comment by praiskup** [comment]
@keszybz would you mind testing Mock with [this patch](https://github.com/praiskup/mock/commit/praiskup-reproduce-copr-2906.patch) with prepared break point?

```
$ rpm -q podman systemd-container mock mock-core-configs kernel-core-6.4.14-200.fc38.x86_64
podman-4.6.2-1.fc38.x86_64
systemd-container-253.9-1.fc38.x86_64
mock-5.0-1.fc38.noarch
mock-core-configs-39-1.fc38.noarch
kernel-core-6.4.14-200.fc38.x86_64
$ rpm -V mock mock-core-configs # no change...
$ wget https://github.com/praiskup/mock/commit/praiskup-reproduce-copr-2906.patch
$ patch --dry-run /usr/lib/python3.11/site-packages/mockbuild/util.py < praiskup-reproduce-copr-2906.patch
$ mock -r fedora-38-x86_64 --scrub=all
$ mock -r fedora-38-x86_64 --shell
...
/usr/bin/systemd-nspawn -q -M 8c78915e41cd46748e298c3465e03026 -D /var/lib/mock/fedora-38-x86_64-bootstrap/root -a --console=pipe --setenv=TERM=vt100 --setenv=SHELL=/bin/bash --setenv=HOME=/var/lib/mock/fedora-38-x86_64-bootstrap/root/installation-homedir --setenv=HOSTNAME=mock --setenv=PATH=/usr/bin:/bin:/usr/sbin:/sbin '--setenv=PROMPT_COMMAND=printf "\033]0;<mock-chroot>\007"' '--setenv=PS1=<mock-chroot> \s-\v\$ ' --setenv=LANG=C.UTF-8 --setenv=LC_MESSAGES=C.UTF-8 --resolv-conf=off /usr/bin/dnf-3 --releasever 38 --setopt=deltarpm=False --setopt=allow_vendor_change=yes --allowerasing --disableplugin=local --disableplugin=spacewalk --disableplugin=versionlock install python3-dnf python3-dnf-plugins-core --setopt=tsflags=nocontexts
...
(Pdb)
```

Mock has a separate mount namespace (some mount points are not visible on host directly) but it has no effect on the `systemd-nspawn` run -- no matter if you run it in the mock process (via `subprocess.call(["bash"])`) or in a separate window (sudo), you should get the error

`Failed to mount /proc/sys (type n/a) on /proc/sys (MS_BIND ""): Invalid argument`

**Comment by praiskup** [comment]
Ok, `/proc` must be a mountpoint to break systemd-nspawn:
```
$ sudo dnf-3 install --disablerepo='*' --setopt releasever=38 --enablerepo fedora --installroot /tmp/newroot bash coreutils
$ sudo mount -t tmpfs -o mode=0755,size=50 test /tmp/newroot/proc/
$ sudo /usr/bin/systemd-nspawn -D /tmp/newroot bash
Spawning container newroot on /tmp/newroot.
Press Ctrl-] three times within 1s to kill container.
Failed to mount /proc/sys (type n/a) on /proc/sys (MS_BIND ""): No such file or directory
Failed to register machine: Machine 'newroot' already exists
```

**Comment by praiskup** [comment]
The Mock fix https://github.com/rpm-software-management/mock/pull/1214 should allow us to not pre-create /proc directory when systemd-nspawn is in use.

**Comment by praiskup** [comment]
@frostyx can you please review the mock PR?

**Comment by praiskup** [comment]
the systemd nspawn problem is moved here https://github.com/systemd/systemd/issues/29174 -> copr part is fixed by mock 5.1
